File: .agents/auditor_m4_1/validate_skills.js
const fs = require('fs');
const path = require('path');

const skillsDir = 'C:\\Users\\Dell\\Documents\\Projects ZK Nexus\\.agents\\skills';
const results = [];

function log(check, status, details) {
  results.push({ check, status, details });
  console.log(`[${status}] ${check}: ${details}`);
}

console.log('=== SKILL PACKAGE FORENSIC INTEGRITY AUDIT ===');

const skills = fs.readdirSync(skillsDir).filter(f => fs.statSync(path.join(skillsDir, f)).isDirectory());
console.log(`Found ${skills.length} skill package folders: ${skills.join(', ')}`);

if (skills.length !== 5) {
  log('Skill Package Count', 'FAIL', `Expected 5 skills, found ${skills.length}`);
} else {
  log('Skill Package Count', 'PASS', 'Found exactly 5 skill packages');
}

const expectedSkills = ['antigravity-agent-manager', 'apify-lead-generation', 'brain-to-docs', 'cold-email', 'ui-ux-pro-max-skill'];
const missing = expectedSkills.filter(s => !skills.includes(s));
if (missing.length > 0) {
  log('Required Skills Presence', 'FAIL', `Missing skills: ${missing.join(', ')}`);
} else {
  log('Required Skills Presence', 'PASS', 'All 5 expected skill packages are present');
}

const prohibitedPatterns = [
  /\bTODO\b/i,
  /\bFIXME\b/i,
  /Lorem ipsum/i,
  /\bplaceholder\b/i,
  /\bdummy\b/i,
  /\bstub\b/i,
  /\bfake\b/i
];

for (const skill of skills) {
  const skillPath = path.join(skillsDir, skill);
  const skillMdPath = path.join(skillPath, 'SKILL.md');

  if (!fs.existsSync(skillMdPath)) {
    log(`${skill} - SKILL.md Existence`, 'FAIL', 'SKILL.md missing');
    continue;
  }

  const content = fs.readFileSync(skillMdPath, 'utf8');

  // 1. YAML Frontmatter Check
  if (!content.startsWith('---')) {
    log(`${skill} - YAML Frontmatter Start`, 'FAIL', 'SKILL.md does not start with ---');
  } else {
    const parts = content.split('---');
    if (parts.length < 3) {
      log(`${skill} - YAML Frontmatter Closure`, 'FAIL', 'Frontmatter closing --- not found');
    } else {
      const frontmatter = parts[1].trim();
      const nameMatch = frontmatter.match(/^name:\s*(.+)$/m);
      const descMatch = frontmatter.match(/^description:\s*(.+)$/m);

      if (!nameMatch) {
        log(`${skill} - Frontmatter Name Field`, 'FAIL', 'name field missing in frontmatter');
      } else if (nameMatch[1].trim() !== skill) {
        log(`${skill} - Frontmatter Name Field`, 'FAIL', `Name mismatch: expected '${skill}', found '${nameMatch[1].trim()}'`);
      } else {
        log(`${skill} - Frontmatter Name Field`, 'PASS', `Name matches '${skill}'`);
      }

      if (!descMatch) {
        log(`${skill} - Frontmatter Description Field`, 'FAIL', 'description field missing');
      } else if (descMatch[1].trim().length < 20) {
        log(`${skill} - Frontmatter Description Field`, 'FAIL', `Description too short (${descMatch[1].trim().length} chars)`);
      } else {
        log(`${skill} - Frontmatter Description Field`, 'PASS', `Description valid (${descMatch[1].trim().length} chars)`);
      }
    }
  }

  // 2. Instruction Depth Check
  const words = content.split(/\s+/).filter(Boolean);
  const lines = content.split(/\r?\n/);
  if (words.length < 200) {
    log(`${skill} - Instruction Depth`, 'FAIL', `Word count too low (${words.length} words)`);
  } else {
    log(`${skill} - Instruction Depth`, 'PASS', `Substantial depth (${words.length} words, ${lines.length} lines)`);
  }

  // 3. Prohibited / Dummy / Stub Pattern Check
  const foundProhibited = [];
  for (const pat of prohibitedPatterns) {
    const match = content.match(pat);
    if (match) {
      foundProhibited.push(match[0]);
    }
  }
  if (foundProhibited.length > 0) {
    log(`${skill} - Authenticity Check`, 'FAIL', `Prohibited terms found: ${foundProhibited.join(', ')}`);
  } else {
    log(`${skill} - Authenticity Check`, 'PASS', 'No dummy, stub, or placeholder terms found');
  }

  // 4. JSON Config / Template Validation
  function scanDir(dir) {
    let jsonFiles = [];
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        jsonFiles = jsonFiles.concat(scanDir(fullPath));
      } else if (entry.isFile() && entry.name.endsWith('.json')) {
        jsonFiles.push(fullPath);
      }
    }
    return jsonFiles;
  }

  const jsonFiles = scanDir(skillPath);
  if (jsonFiles.length === 0) {
    log(`${skill} - Config Files`, 'WARN', 'No JSON config files found');
  } else {
    for (const jFile of jsonFiles) {
      const relPath = path.relative(skillPath, jFile);
      try {
        const jRaw = fs.readFileSync(jFile, 'utf8');
        const jParsed = JSON.parse(jRaw);
        
        // Authenticity check on JSON
        const jProhibited = [];
        for (const pat of prohibitedPatterns) {
          const match = jRaw.match(pat);
          if (match) jProhibited.push(match[0]);
        }

        if (jProhibited.length > 0) {
          log(`${skill} - ${relPath} Authenticity`, 'FAIL', `Prohibited terms in JSON: ${jProhibited.join(', ')}`);
        } else {
          log(`${skill} - ${relPath} JSON Validity & Authenticity`, 'PASS', `Valid JSON syntax and authentic content`);
        }
      } catch (err) {
        log(`${skill} - ${relPath} JSON Syntax`, 'FAIL', `JSON parse error: ${err.message}`);
      }
    }
  }
}

console.log('\n=======================================');
const failures = results.filter(r => r.status === 'FAIL');
if (failures.length > 0) {
  console.log(`FINAL AUDIT VERDICT: INTEGRITY VIOLATION (${failures.length} check failures)`);
  process.exit(1);
} else {
  console.log('FINAL AUDIT VERDICT: CLEAN (All integrity, syntax, and depth checks passed)');
  process.exit(0);
}

File: .agents/auditor_m4_1/validate_skills.py
import os
import json
import re
import sys

skills_dir = r"C:\Users\Dell\Documents\Projects ZK Nexus\.agents\skills"
audit_results = []

def log_result(check_name, status, details):
    audit_results.append({
        "check": check_name,
        "status": status,
        "details": details
    })
    print(f"[{status}] {check_name}: {details}")

print("=== Starting Skill Package Integrity Verification ===")

# Discover skills
skills = [d for d in os.listdir(skills_dir) if os.path.isdir(os.path.join(skills_dir, d))]
print(f"Found {len(skills)} skill packages: {skills}")

if len(skills) != 5:
    log_result("Skill Package Count Check", "FAIL", f"Expected 5 skill packages, found {len(skills)}")
else:
    log_result("Skill Package Count Check", "PASS", f"Found exactly 5 skill packages: {skills}")

expected_skills = ["antigravity-agent-manager", "apify-lead-generation", "brain-to-docs", "cold-email", "ui-ux-pro-max-skill"]
missing_skills = [s for s in expected_skills if s not in skills]
if missing_skills:
    log_result("Required Skills Check", "FAIL", f"Missing expected skills: {missing_skills}")
else:
    log_result("Required Skills Check", "PASS", "All required skill packages exist.")

# Audit each skill package
for skill in skills:
    skill_path = os.path.join(skills_dir, skill)
    skill_md = os.path.join(skill_path, "SKILL.md")
    
    # Check SKILL.md existence
    if not os.path.exists(skill_md):
        log_result(f"{skill} SKILL.md Existence", "FAIL", "SKILL.md missing")
        continue
    
    with open(skill_md, "r", encoding="utf-8") as f:
        content = f.read()
    
    # 1. Frontmatter check
    if not content.startswith("---"):
        log_result(f"{skill} YAML Frontmatter Header", "FAIL", "SKILL.md does not start with ---")
    else:
        parts = content.split("---", 2)
        if len(parts) < 3:
            log_result(f"{skill} YAML Frontmatter Structure", "FAIL", "SKILL.md frontmatter closing --- not found")
        else:
            frontmatter_text = parts[1].strip()
            # Simple YAML parsing check
            name_match = re.search(r"^name:\s*(.+)$", frontmatter_text, re.MULTILINE)
            desc_match = re.search(r"^description:\s*(.+)$", frontmatter_text, re.MULTILINE)
            
            if not name_match:
                log_result(f"{skill} Frontmatter 'name'", "FAIL", "Missing 'name' field in frontmatter")
            elif name_match.group(1).strip() != skill:
                log_result(f"{skill} Frontmatter 'name'", "FAIL", f"Name mismatch: expected '{skill}', got '{name_match.group(1).strip()}'")
            else:
                log_result(f"{skill} Frontmatter 'name'", "PASS", f"Name correctly matches '{skill}'")
                
            if not desc_match:
                log_result(f"{skill} Frontmatter 'description'", "FAIL", "Missing 'description' field in frontmatter")
            elif len(desc_match.group(1).strip()) < 20:
                log_result(f"{skill} Frontmatter 'description'", "FAIL", "Description is too short (< 20 chars)")
            else:
                log_result(f"{skill} Frontmatter 'description'", "PASS", f"Description valid ({len(desc_match.group(1).strip())} chars)")

    # 2. Instruction depth & quality check
    lines = content.splitlines()
    words = content.split()
    if len(words) < 200:
        log_result(f"{skill} Content Depth", "FAIL", f"Content depth insufficient ({len(words)} words)")
    else:
        log_result(f"{skill} Content Depth", "PASS", f"Sufficient content depth ({len(words)} words, {len(lines)} lines)")

    # 3. Dummy / Stub / Fake text check
    suspicious_patterns = [r"\bTODO\b", r"\bFIXME\b", r"Lorem ipsum", r"placeholder", r"dummy", r"stub", r"fake"]
    found_suspicious = []
    for pattern in suspicious_patterns:
        matches = re.findall(pattern, content, re.IGNORECASE)
        if matches:
            found_suspicious.append(f"{pattern} ({len(matches)} matches)")
            
    if found_suspicious:
        log_result(f"{skill} Authenticity Check", "FAIL", f"Found prohibited/suspicious patterns: {', '.join(found_suspicious)}")
    else:
        log_result(f"{skill} Authenticity Check", "PASS", "No dummy/stub/placeholder patterns found in SKILL.md.")

    # 4. JSON Config / Template Files Validation
    json_files = []
    for root, _, files in os.walk(skill_path):
        for file in files:
            if file.endswith(".json"):
                json_files.append(os.path.join(root, file))
                
    if not json_files:
        log_result(f"{skill} Config/Template Files", "WARN", "No JSON config or template files found.")
    else:
        for jf in json_files:
            rel_j_path = os.path.relpath(jf, skill_path)
            try:
                with open(jf, "r", encoding="utf-8") as jf_in:
                    j_data = json.load(jf_in)
                
                # Check for suspicious text in json
                j_str = json.dumps(j_data)
                found_j_suspicious = []
                for pattern in suspicious_patterns:
                    matches = re.findall(pattern, j_str, re.IGNORECASE)
                    if matches:
                        found_j_suspicious.append(f"{pattern} ({len(matches)} matches)")
                
                if found_j_suspicious:
                    log_result(f"{skill} Config {rel_j_path} Authenticity", "FAIL", f"Prohibited text in JSON: {', '.join(found_j_suspicious)}")
                else:
                    log_result(f"{skill} Config {rel_j_path} Syntax & Authenticity", "PASS", f"Valid JSON syntax and authentic content in {rel_j_path}")
            except Exception as e:
                log_result(f"{skill} Config {rel_j_path} JSON Syntax", "FAIL", f"Invalid JSON syntax: {e}")

# Summary report
print("\n=== VERDICT SUMMARY ===")
fails = [r for r in audit_results if r["status"] == "FAIL"]
if fails:
    print(f"VERDICT: INTEGRITY VIOLATION ({len(fails)} failures)")
else:
    print("VERDICT: CLEAN (All integrity and syntax checks passed)")
